nmigen/vendor/lattice_ice40.py

```diff
@@ -350,14 +361,39 @@ def create_missing_domain(self, name):
         # The power-on reset timer counts up because the vendor tools do not support initialization
         # of flip-flops.
         if name == "sync" and self.default_clk is not None:
-            clk_i = self.request(self.default_clk).i
+            m = Module()
+
+            # Internal high-speed clock: 6MHz, 12MHz, 24MHz, or 48MHz depending on the divider.
+            if self.default_clk == "SB_HFOSC":
+                if not hasattr(self, "hfosc_div"):
+                    raise ValueError("SB_HFOSC divider exponent (hfosc_div) must be an integer "
+                                     "between 0 and 3")
+                if not isinstance(self.hfosc_div, int) or self.hfosc_div < 0 or self.hfosc_div > 3:
+                    raise ValueError("SB_HFOSC divider exponent (hfosc_div) must be an integer "
+                                     "between 0 and 3, not {!r}"
+                                     .format(self.hfosc_div))
+                clk_i = Signal()
+                m.submodules += Instance("SB_HFOSC",
+                                         i_CLKHFEN=1,
+                                         i_CLKHFPU=1,
+                                         p_CLKHF_DIV="0b{0:b}".format(self.hfosc_div),
+                                         o_CLKHF=clk_i)
+            # Internal low-speed clock: 10KHz.
+            elif self.default_clk == "SB_LFOSC":
+                clk_i = Signal()
+                m.submodules += Instance("SB_LFOSC",
+                                         i_CLKLFEN=1,
+                                         i_CLKLFPU=1,
+                                         o_CLKLF=clk_i)
+            # User-defined clock signal.
+            else:
+                clk_i = self.request(self.default_clk).i
+
             if self.default_rst is not None:
                 rst_i = self.request(self.default_rst).i
             else:
                 rst_i = Const(0)
 
-            m = Module()
-
             # Power-on-reset domain
             m.domains += ClockDomain("por", reset_less=True, local=True)
             delay = int(15e-6 * self.default_clk_frequency)
```

Comment:
Both SB_HFOSC and SB_LFOSC output is unstable for 100 µs after enabling them. So if one of these oscillators are selected, this power-on reset delay should be increased to 100 µs here.

Reply:
Huh, you learn something new every day. Should it have a 5x margin like the BRAM delay and be 500us, or is 100us okay?

Reply:
Nope, 100 us is OK. The reason BRAM delay has a 5x margin is that the 3 us figure is experimental (i.e. typical case) rather than provided by the vendor (i.e. worst case). Since we have no idea where that delay comes from it seems wise to play it safe, but that's not needed here.